pkg/logical-cluster-manager/cluster-manager-controller.go

```diff
@@ -0,0 +1,244 @@
+	kubeconfig       *string
+	ctx              context.Context
+	clusterclientset edgeclient.Interface
+	clusterInformer  v1alpha1.LogicalClusterInformer
+	queue            workqueue.RateLimitingInterface
+}
+
+func NewController(
+	kubeconfig *string,
+	ctx context.Context,
+	clusterclientset edgeclient.Interface,
+	clusterInformer v1alpha1.LogicalClusterInformer) *Controller {
+	logger := klog.FromContext(ctx)
+
+	// TODO: We are keeping a hash table of logical cluster provider clients
+	// this might change once we create the cluster provider client resource.
+	clusterproviderclient.ProviderList = make(map[string]clusterproviderclient.ProviderClient)
+
+	controller := &Controller{
+		kubeconfig:       kubeconfig,
+		ctx:              ctx,
+		clusterclientset: clusterclientset,
+		clusterInformer:  clusterInformer,
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(),
+			"cluster-controller"),
+	}
+
+	logger.Info("Setting up event handlers")
+	clusterInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			logger.Info("Add cluster")
+			controller.queue.Add(
+				queueItem{
+					op:  opAdd,
+					key: obj,
+				},
+			)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			// Most fields in the logical cluster are immutable with the
+			// exception of the URL and fields in the status.
+			oldInfo := oldObj.(*v1alpha1apis.LogicalCluster)
+			newInfo := newObj.(*v1alpha1apis.LogicalCluster)
+			if !reflect.DeepEqual(oldInfo.Status, newInfo.Status) {
+				controller.queue.Add(
+					queueItem{
+						op:  opUpdate,
+						key: newObj,
+					},
+				)
+			}
+		},
+		DeleteFunc: func(delObj interface{}) {
+			logger.Info("Delete cluster")
+			controller.queue.Add(
+				queueItem{
+					op:  opDelete,
+					key: delObj,
+				},
+			)
+		},
+	})
+
+	logger.Info("New %s controller", controllerName)
+	return controller
+}
+
+// Run starts the controller, which stops when c.context.Done() is closed.
+func (c *Controller) Run(numThreads int) {
+	defer runtime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	logger := klog.FromContext(c.ctx)
+	logger.Info("starting controller")
+	defer logger.Info("shutting down controller")
+
+	for i := 0; i < numThreads; i++ {
+		go wait.UntilWithContext(c.ctx, c.runWorker, time.Second)
+	}
+
+	<-c.ctx.Done()
+}
+
+func (c *Controller) runWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	i, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(i)
+
+	//ES: not sure we should treat all errors in the same way (queue back)
+	if err := c.process(ctx, i.(queueItem)); err != nil {
+		runtime.HandleError(fmt.Errorf("%q controller didn't sync, err: %w", controllerName, err))
```

Comment:
A retry is not an un-handled error.

Also: https://kubernetes.slack.com/archives/C0EG7JC6T/p1686033859701419